File: Classes/ThemeObject/UIButton+LWTheme.m
//
//  UIButton+LWTheme.m
//  LWTheme
//
//  Created by lalawue on 2019/12/11.
//  Copyright © 2019 lalawue. All rights reserved.
//

#import "UIButton+LWTheme.h"
#import "LWTheme_Tools.h"

@interface LWButtonImageColorItem : NSObject
@property (nonatomic, strong) UIImage *image;
@property (nonatomic, strong) UIImage *backgroundImage;
@property (nonatomic, copy) NSString *imageTintColorKey;
@property (nonatomic, copy) NSString *backgroundImageTintColorKey;
@end

@implementation LWButtonImageColorItem
@end

#pragma mark -

@implementation UIButton (LWTheme)

- (void)lw_setTitleColor:(NSString *)colorPattern forState:(UIControlState)state
{
    NSMutableDictionary *dict = [self lw_getTitleColorDict] ?: [NSMutableDictionary dictionaryWithCapacity:2];
    if (colorPattern) {
        [dict setObject:colorPattern forKey:@(state)];
    } else {
        [dict removeObjectForKey:@(state)];
    }
    UIColor *colorValue = [self lw_colorForPattern:colorPattern];
    if (colorValue) {
        [self setTitleColor:colorValue forState:state];
    }
    [self lw_storeValue:dict forSetter:@selector(lw_setTitleColorDict:)];
}

- (NSString *)lw_titleColorForState:(UIControlState)state
{
    NSDictionary *dict = [self lw_getTitleColorDict];
    return [dict objectForKey:@(state)];
}

- (void)lw_setImage:(UIImage *)image withTintColor:(NSString *)tintColor forState:(UIControlState)state
{
    NSNumber *stateKey = @(state);
    NSMutableDictionary *dict = [self lw_getImageStateDict] ?: [NSMutableDictionary dictionaryWithCapacity:2];
    LWButtonImageColorItem *item = dict[stateKey] ?: [LWButtonImageColorItem new];
    
    item.image = image;
    item.imageTintColorKey = tintColor;
    
    dict[stateKey] = item;
    
    if (image) {
        UIColor *color = [self lw_colorForPattern:tintColor];
        image = [LWImageUtils fillImage:image withColor:color];
    }
    [self setImage:image forState:state];
    [self lw_storeValue:dict forSetter:@selector(lw_setImageForStateDict:)];
}

- (void)lw_setBackgroundImage:(UIImage *)image withTintColor:(NSString *)tintColor forState:(UIControlState)state
{
    NSNumber *stateKey = @(state);
    NSMutableDictionary *dict = [self lw_getImageStateDict] ?: [NSMutableDictionary dictionaryWithCapacity:2];
    LWButtonImageColorItem *item = dict[stateKey] ?: [LWButtonImageColorItem new];
    
    item.backgroundImage = image;
    item.backgroundImageTintColorKey = tintColor;

    dict[stateKey] = item;

    if (image) {
        UIColor *color = [self lw_colorForPattern:tintColor];
        image = [LWImageUtils fillImage:image withColor:color];
    }
    [self setBackgroundImage:image forState:state];
    [self lw_storeValue:dict forSetter:@selector(lw_setImageForStateDict:)];
}

#pragma mark - Apply

// title color
- (void)lw_setTitleColorDict:(NSMutableDictionary *)dict
{
    NSArray *array = [dict.allKeys copy];
    for (NSNumber *stateKey in array) {
        NSString *colorPattern = dict[stateKey];
        UIColor *colorValue = [self lw_colorForPattern:colorPattern];
        if (colorValue) {
            UIControlState state = [stateKey unsignedIntegerValue];
            [self setTitleColor:colorValue forState:state];
        }
    }
}

- (NSMutableDictionary *)lw_getTitleColorDict
{
    return (NSMutableDictionary *)[self lw_valueForSetter:@selector(lw_setTitleColorDict:)];
}

// image tint color state
- (void)lw_setImageForStateDict:(NSDictionary *)dict
{
    NSArray *array = [dict.allKeys copy];
    for (NSNumber *stateKey in array) {
        UIControlState state = [stateKey unsignedIntegerValue];
        LWButtonImageColorItem *item = dict[stateKey];
        // set tint image
        if (item.image && item.imageTintColorKey.length) {
            UIColor *colorValue = [self lw_colorForPattern:item.imageTintColorKey];
            [self setImage:[LWImageUtils fillImage:item.image withColor:colorValue]
                  forState:state];
        }
        // set tint background image
        if (item.backgroundImage && item.backgroundImageTintColorKey.length) {
            UIColor *colorValue = [self lw_colorForPattern:item.backgroundImageTintColorKey];
            [self setBackgroundImage:[LWImageUtils fillImage:item.backgroundImage withColor:colorValue]
                            forState:state];
        }
    }
}

- (NSMutableDictionary *)lw_getImageStateDict
{
    return (NSMutableDictionary *)[self lw_valueForSetter:@selector(lw_setImageForStateDict:)];
}

@end

File: Classes/ThemeObject/UIButton+LWTheme.swift
//
//  UIButton+LWTheme.swift
//  LWThemeSwift
//
//  Created by lalawue on 2019/12/18.
//  Copyright © 2019 lalawue. All rights reserved.
//

import UIKit

fileprivate class ImageTintColorItem : NSObject {
    var image : UIImage? = nil
    var backgroundImage : UIImage? = nil
    var imageTintColorKey : String? = nil
    var backgroundImageTintColorKey : String? = nil
}

extension UIButton {
    
    @objc public func lw_setTitleColor(_ colorPattern: String?, for state: UIControl.State) {
        let stateKey = NSNumber(value: state.rawValue)
        var dict = lw_getTitleColorDict() ?? [:]

        if (colorPattern != nil) {
            dict[stateKey] = colorPattern!
        } else {
            dict[stateKey] = nil
        }
        
        if (colorPattern != nil) {
            let color = lw_colorForPattern(colorPattern)
            setTitleColor(color, for: state)
        }

        lw_storeValue(dict, setter: #selector(lw_setTitleColorDict(_:)))
    }
    
    @objc public func lw_titleColor(for state: UIControl.State) -> String? {
        return lw_getTitleColorDict()?[NSNumber(value: state.rawValue)]
    }
    
    @objc public func lw_setImage(_ image: UIImage?, tintColor : String?, for state: UIControl.State) {
        let stateKey = NSNumber(value: state.rawValue)
        var dict = lw_getImageStateDict() ?? [:]
        let item = dict[stateKey] ?? ImageTintColorItem()
        
        item.image = image
        item.imageTintColorKey = tintColor

        dict[stateKey] = item
        
        if (tintColor != nil) {
            let color = lw_colorForPattern(tintColor)
            let imageValue = image?.lw_tintImageWithColor(color, blendMode: CGBlendMode.destinationIn)
            setImage(imageValue, for: state)
        }
        
        lw_storeValue(dict, setter: #selector(UIButton.lw_setImageStateDict(_:)))
    }
    
    
    @objc public func lw_setBackgroundImage(_ image: UIImage?, tintColor : String?, for state: UIControl.State) {
        let stateKey = NSNumber(value: state.rawValue)
        var dict = lw_getImageStateDict() ?? [:]
        let item = dict[stateKey] ?? ImageTintColorItem()
        
        item.backgroundImage = image
        item.backgroundImageTintColorKey = tintColor
        
        dict[stateKey] = item
        
        if (tintColor != nil) {
            let color = lw_colorForPattern(tintColor)
            let imageValue = image?.lw_tintImageWithColor(color, blendMode: CGBlendMode.destinationIn)
            setBackgroundImage(imageValue, for: state)
        }
        
        lw_storeValue(dict, setter: #selector(UIButton.lw_setImageStateDict(_:)))
    }
    
    //MARK: Apply
    
    // title color
    fileprivate func lw_getTitleColorDict() -> [NSNumber : String]? {
        return lw_valueForSetter(#selector(UIButton.lw_setTitleColorDict(_:))) as? [NSNumber : String]
    }
    
    @objc fileprivate func lw_setTitleColorDict(_ dict : [NSNumber : String]) {
        for (stateKey, colorPattern) in dict {
            let state = UIControl.State.init(rawValue: stateKey.uintValue)
            let colorValue = lw_colorForPattern(colorPattern)
            setTitleColor(colorValue, for: state)
        }
    }
    
    // image & bacgroundImage tint color for state
    fileprivate func lw_getImageStateDict() -> [NSNumber : ImageTintColorItem]? {
        return lw_valueForSetter(#selector(UIButton.lw_setImageStateDict(_:))) as? [NSNumber : ImageTintColorItem]
    }
    
    @objc fileprivate func lw_setImageStateDict(_ dict : [NSNumber : ImageTintColorItem]) {
        for (stateKey, item) in dict {
            // set image
            if ((item.image != nil) && (item.imageTintColorKey != nil)) {
                let colorValue = lw_colorForPattern(item.imageTintColorKey)
                if (colorValue != nil) {
                    let state = UIControl.State.init(rawValue: stateKey.uintValue)
                    let image = item.image?.lw_tintImageWithColor(colorValue!, blendMode: CGBlendMode.destinationIn)
                    setImage(image, for: state)
                }
            }
            // set background image
            if ((item.backgroundImage != nil) && (item.backgroundImageTintColorKey != nil)) {
                let colorValue = lw_colorForPattern(item.backgroundImageTintColorKey)
                if (colorValue != nil) {
                    let state = UIControl.State.init(rawValue: stateKey.uintValue)
                    let image = item.backgroundImage?.lw_tintImageWithColor(colorValue!, blendMode: CGBlendMode.destinationIn)
                    setBackgroundImage(image, for: state)
                }
            }
        }
    }
}
